runtime/lua/vim/treesitter/_query_linter.lua

```diff
@@ -0,0 +1,182 @@
+local namespace = vim.api.nvim_create_namespace('vim.treesitter.query_linter')
+-- those node names exist for every language
+local BUILT_IN_NODE_NAMES = { '_', 'ERROR' }
+
+local M = {}
+
+---@private
+--- Adds a diagnostic for node in the query buffer
+local function add_lint_for_node(diagnostics, node, buf, lint, lang, severity)
+  local node_text = vim.treesitter.get_node_text(node, buf):gsub('\n', ' ')
+  local message = lint .. ': ' .. node_text
+  local error_range = { node:range() }
+  table.insert(diagnostics, {
+    lnum = error_range[1],
+    end_lnum = error_range[3],
+    col = error_range[2],
+    end_col = error_range[4],
+    severity = severity or vim.diagnostic.ERROR,
+    message = message,
+    source = lang,
+  })
+end
+
+---@private
+--- like vim.tbl_contains but works with function predicate
+--- TODO: Should be handled by standard vim.* functions!
+--- vim.tbl_contains can not be used because it does not do nested comparison (like vim.deep_equal)
+--- and the table returned by vim.treesitter.language.inspect is a mixed list/dictionary-like table
+local function table_contains(predicate, table)
+  for _, elt in pairs(table) do
+    if predicate(elt) then
+      return true
+    end
+  end
+  return false
+end
+
+---@private
+--- Determines the target language of a query file by its path: <lang>/<query_type>.scm
+local function guess_query_lang(buf, _)
+  local filename = vim.api.nvim_buf_get_name(buf)
+  if filename ~= '' then
+    local ok, query_lang = pcall(vim.fn.fnamemodify, filename, ':p:h:t')
+    return ok and query_lang
+  end
+end
+
+---@private
+local function normalize_opts(buf, opts)
+  opts = opts or {}
+  opts.langs = opts.langs or guess_query_lang
+  if type(opts.langs) == 'function' then
+    opts.langs = opts.langs(buf, opts)
+  end
+  if type(opts.langs) ~= 'table' then
+    opts.langs = { opts.langs }
+  end
+  return opts
+end
+
+local lint_query = [[;; query
+ ; TODO: this check is still costly. Enable this when performance issues are addressed
+ ; (program [(named_node) (list) (grouping)] @toplevel)
```

Comment:
I suggest to disable this costly check for now, when query linter should run on TextChanged/InsertLeave. Without the costly check it feels snappier. However, the costly check is the final ground truth whether Neovim will accept the query or not.

Leaving it in this draft PR for reviewers to experiment.